contributors/design-proposals/auditing.md

```diff
@@ -0,0 +1,290 @@
+
+### Events
+
+```go
+package audit
+
+// AuditLevel defines the amount of information logged during auditing
+type AuditLevel string
+
+// Valid audit levels
+const (
+    // AuditHttpHeader provides basic level of auditing, logging data at HTTP level
+    AuditHttpHeader AuditLevel   = "HttpHeader"
+    // AuditRequestobject provides basic level of auditing, and additionally
+    // logs unstructured request object
+    AuditRequestObject AuditLevel = "RequestObject"
+    // AuditStorageobject provides basic and request level, and additionally
+    // logs object before and after saving in storage
+    AuditStorageObject AuditLevel = "StorageObject"
+)
+
+type Event struct {
+    // AuditLevel at which event was generated
+    Level AuditLevel
+
+    // below fields are filled at HttpHeader level and higher:
+
+    // Unique audit ID, generated for each request
+    ID string
+    // Time the event reached the apiserver
+    Timestamp Timestamp
+    // Source IP, from where the request originates
+    SourceIP string
+    // HTTP method sent by the client
+    HttpMethod string
+    // Authentication method used to allow users access the cluster
+    AuthMethod string
+    // RequestURI is the Request-Line as sent by the client to a server
+    RequestURI string
+    // User information
+    User UserInfo
+    // Impersonate-User header value
+    AsUser string
+    // Impersonate-Group header value
+    AsGroups string
+    // Object reference this request is targeted at
+    Object ObjectReference
+    // Response status code are returned by the server
+    ResponseStatusCode int
+
+    // below fields are filled at RequestObject level and higher:
+
+    // RequestObject logged before admission
+    RequestObject runtime.Unstructured
+
+    // below fields are filled at StorageObject level and higher:
+
+    // Object value before modification (will be empty when creating new object)
+    OldObject runtime.Object
+    // Object value after modification (will be empty when removing object)
+    NewObject runtime.Object
+    // Difference between old and new object
+    Patch     []byte
+}
+```
+
+### Concurrency/Transaction Model for Events
+
+TODO: define (and maybe adapt interface) how synchronization of the writable audit event of a context takes place.
+
+### Levels
+
+```go
+type Action interface {
+  Level() LogLevel
+  Value(key string) string
+}
+
+type Policy interface {
+  func Action(kind, namespace, method, user string) Action
+}
+```
+
+TODO: how do the storage layer and the http handlers get access to the policy?
+
+### Filters
+
+TODO: define what we need next to the audit level
+
+### Output Backend Interface
+
+```go
+package audit
+
+type OutputBackend interface {
```

Comment:
Some details worth expanding on:
- Is `OutputBackend.Log` expected to be blocking or non-blocking?
- Where should things like buffering & retry logic be implemented? (probably the repsonsibility of the OutputBackend?)
- How are errors handled? (attempt to log to local apiserver log & expose a prometheus metric)